scrapy_poet/spidermiddlewares.py

```diff
@@ -0,0 +1,31 @@
+try:
+    from web_poet.exceptions import Retry
+except ImportError:
+    pass
+else:
+    from typing import List, Optional
+
+    from scrapy import Spider
+    from scrapy.downloadermiddlewares.retry import get_retry_request
+    from scrapy.http import Request, Response
+
+    class RetrySpiderMiddleware:
+        """Captures :exc:`web_poet.exceptions.Retry` exceptions from spider
+        callbacks, and retries the source request."""
+
+        def process_spider_exception(
+            self,
+            response: Response,
+            exception: BaseException,
+            spider: Spider,
+        ) -> Optional[List[Request]]:
+            if not isinstance(exception, Retry):
+                return None
+            new_request_or_none = get_retry_request(
+                response.request,
+                spider=spider,
+                reason="page_object_retry",
+            )
+            if not new_request_or_none:
+                return []
+            return [new_request_or_none]
```

Comment:
Funny that we are getting this feature here before upstream Scrapy gets it.

I thought about eventually relying on the Scrapy implementation, e.g. having a `RETRY_EXCEPTIONS` Scrapy setting and extending it with `web_poet.exceptions.Retry` as part of the web-poet configuration, instead of installing a middleware. However, the handling of the retry `reason` I think makes that complicated.

So I think the way to go will be to eventually have projects with both, a `RetrySpiderMiddleware` enabled by Scrapy by default that captures a future Scrapy `Retry` exception that supports a `reason` parameter, and this other `RetrySpiderMiddleware` specifically for web-poet support.

Reply:
I'm thinking if `web_poet.exceptions.Retry` should accept `**kwargs` params. This is quite similar to https://github.com/psf/requests/blob/main/requests/exceptions.py#L17. In this way, it doesn't blow up with `TypeError: Retry() takes no keyword arguments`.

In any case, I'm good with having 2 separate middlewares at the moment, just in case the retry behaviors _might_ diverge in the future. 👍